### PR TITLE
[relay][external codegen] outline and inline lifted functions for external codegen

### PR DESCRIPTION
**Pull request by zhiics**
This PR outlines the functions that will be handled by external codegen and then inline them back to avoid the optimizations from Relay. All function level passes are handled directly from the pass manager. A few module level passes needs to detect if a function uses default compiler or not. This is because these passes take the functions internally and they are invisible to the pass manager.

### CODE REVIEW

**Comment by zhiics** [comment]
cc @comaniac @mbaret @soiferj @tqchen @icemelon9 @masahi

File: src/relay/pass/partition_graph.cc [hunk 1]
```diff
@@ -207,16 +211,24 @@ class Partitioner : public ExprMutator {
...
       std::string name = compiler_attrs->compiler + "_" + std::to_string(subgraph->id);
       subgraph_func =
           FunctionSetAttr(subgraph_func, attr::kExternalSymbol, tir::StringImmNode::make(name));
       subgraph_func = FunctionSetAttr(subgraph_func, attr::kPrimitive, tvm::Integer(1));
       subgraph_func = FunctionSetAttr(subgraph_func, attr::kCompiler,
                                       tvm::tir::StringImmNode::make(compiler_attrs->compiler));
-      return CallNode::make(subgraph_func, args);
+      subgraph_func = FunctionSetAttr(subgraph_func, attr::kInline, tvm::Integer(1));
+      CHECK(!module_->ContainGlobalVar(name))
+          << "Global function " << name << " is already existed";
```
Comment (mbaret) [inline src/relay/pass/partition_graph.cc]: " already exists"

File: src/relay/backend/vm/inline_primitives.cc [hunk 2]
```diff
@@ -122,6 +122,7 @@ struct PrimitiveInliner : ExprMutator {
       auto global = pair.first;
       auto base_func = pair.second;
       if (auto* n = base_func.as<FunctionNode>()) {
+        if (!n->UseDefaultCompiler()) continue;
```
Comment (mbaret) [inline src/relay/backend/vm/inline_primitives.cc]: Presumably all future passes which operate at the module level will also need to include a similar line to this. Would it be meaningful to mark some global functions as 'do not touch'? That might be more explicit than having to know that you shouldn't edit functions that don't use the default compiler.

**Comment by masahi** [comment]
How about adding batch norm to the test to make sure it is still there after opt passes?

File: src/relay/pass/partition_graph.cc [hunk 3]
```diff
@@ -207,16 +211,24 @@ class Partitioner : public ExprMutator {
...
+      GlobalVar glob_func(name);
```
Comment (comaniac) [inline src/relay/pass/partition_graph.cc]: It might be clearer to add a comment saying we are creating a global (outline) function that won't be touched by other passes until inline.

File: src/relay/backend/vm/inline_primitives.cc (hunk 2 above)
Comment (zhiics) [reply src/relay/backend/vm/inline_primitives.cc]: We probably cannot do that. All functions in prelude are global functions. Also lamb lifting will lift some closures to the global scope. These functions should be treated as normal ones.

**Comment by zhiics** [comment]
@masahi Good point. Done

File: tests/python/relay/test_pass_partition_graph.py [hunk 4]
```diff
@@ -427,10 +465,110 @@ def test_extern_dnnl_mobilenet():
...
+        mod["main"] = func
+        op_list = ["nn.batch_norm", "nn.conv2d"]
+        mod = WhiteListAnnotator(op_list, "test_compiler")(mod)
+
+        opt_pass = transform.Sequential([
+            transform.InferType(),
+            transform.PartitionGraph(),
+            transform.SimplifyInference(),
+            transform.FoldConstant(),
+            transform.AlterOpLayout()
```
Comment (masahi) [inline tests/python/relay/test_pass_partition_graph.py]: I think adding Inline here would better illustrate the motivation of this PR (also closer to `relay.optimize` pipeline).

I get
```
def @main(%data: Tensor[(1, 3, 224, 224), float32], %weight: Tensor[(16, 3, 3, 3), float32], %bn_gamma: Tensor[(16), float32], %bn_beta: Tensor[(16), float32], %bn_mean: Tensor[(16), float32], %bn_var: Tensor[(16), float32]) -> (Tensor[(1, 16, 224, 224), float32], Tensor[(16), float32], Tensor[(16), float32]) {
  %0 = fn (%test_compiler_input0: Tensor[(1, 3, 224, 224), float32], %test_compiler_input1: Tensor[(16, 3, 3, 3), float32], Inline=1, Compiler="test_compiler", ExternalSymbol="test_compiler_1", Primitive=1) -> Tensor[(1, 16, 224, 224), float32] {
    nn.conv2d(%test_compiler_input0, %test_compiler_input1, padding=[1, 1, 1, 1], channels=16, kernel_size=[3, 3]) /* ty=Tensor[(1, 16, 224, 224), float32] */
  };
  %1 = %0(%data, %weight) /* ty=Tensor[(1, 16, 224, 224), float32] */;
  %2 = fn (%test_compiler_input2: Tensor[(1, 16, 224, 224), float32], %test_compiler_input3: Tensor[(16), float32], %test_compiler_input4: Tensor[(16), float32], %test_compiler_input5: Tensor[(16), float32], %test_compiler_input6: Tensor[(16), float32], Inline=1, Compiler="test_compiler", ExternalSymbol="test_compiler_0", Primitive=1) -> (Tensor[(1, 16, 224, 224), float32], Tensor[(16), float32], Tensor[(16), float32]) {
    nn.batch_norm(%test_compiler_input2, %test_compiler_input3, %test_compiler_input4, %test_compiler_input5, %test_compiler_input6) /* ty=(Tensor[(1, 16, 224, 224), float32], Tensor[(16), float32], Tensor[(16), float32]) */
  };
  %2(%1, %bn_gamma, %bn_beta, %bn_mean, %bn_var) /* ty=(Tensor[(1, 16, 224, 224), float32], Tensor[(16), float32], Tensor[(16), float32]) */
}
```
which I think is more interesting

File: src/relay/pass/partition_graph.cc [hunk 5]
```diff
@@ -207,16 +211,28 @@ class Partitioner : public ExprMutator {
...
       subgraph_func = FunctionSetAttr(subgraph_func, attr::kPrimitive, tvm::Integer(1));
       subgraph_func = FunctionSetAttr(subgraph_func, attr::kCompiler,
                                       tvm::tir::StringImmNode::make(compiler_attrs->compiler));
-      return CallNode::make(subgraph_func, args);
+      subgraph_func = FunctionSetAttr(subgraph_func, attr::kInline, tvm::Integer(1));
+      CHECK(!module_->ContainGlobalVar(name))
+          << "Global function " << name << " already exists";
+      // Create a global function and add it to the IRModule for the subgraph.
+      // This way we lift the functions that should be handled by external
+      // codegen to the module scope and rely the pass manager to prevent relay
```
Comment (masahi) [inline src/relay/pass/partition_graph.cc]: rely on

**Comment by masahi** [comment]
Will wait for @mbaret's approval.

File: src/relay/backend/vm/inline_primitives.cc (hunk 2 above)
Comment (mbaret) [reply src/relay/backend/vm/inline_primitives.cc]: Understood.

**Comment by masahi** [comment]
Also need to wait for the github bug to be resolved

File: tests/python/relay/test_pass_partition_graph.py (hunk 4 above)
Comment (zhiics) [reply tests/python/relay/test_pass_partition_graph.py]: Let's keep the previous one as it tests the lifting. We can create a separate one to test batch_norm with inline. There are a few other tests for outlining and inlining tests in this file as well. I think this should be enough.

**Comment by masahi** [comment]
Thanks @zhiics @mbaret @comaniac

**Comment by mbaret** [comment]
@zhiics It looks like this makes it so we can't run any passes on external functions. Is this the intended behaviour, and if so are we intending to add a mechanism at some point to allow for passes on external functions?

**Comment by zhiics** [comment]
Yes, this is intended. We plan to do it. For example, we may want to have a build pipeline that executes target independent passes first and then target dependent passes. We can partition after platform independent passes and rely on external compilers to take care of them.

You can do this right now manually. For example, you can run constant folding, CSE, etc, before partitioning.

**Comment by mbaret** [comment]
Beforehand, we had been running some passes after partitioning (for instance a layout transform for ACL functions). It looks like we can't even force these to run now though as no function passes are allowed to touch external functions. We can't run the passes before partitioning because it will transform the entire graph, not just the bits for external partitioning.

**Comment by zhiics** [comment]
I personally don't think TVM should be aware of accelerator specific transformations because they are very hardware specific. For example, we don't want this on TRT, dnnl, and many other accelerators. However, you can still do this even now. You can directly run layout transformation by yourself.

**Comment by mbaret** [comment]
It's possible we're doing something wrong, but the function LowerExternalFunctions in compile_engine produces modules just containing external functions. We want to call passes on these modules from our ACL-specific codegen (so they'll only affect our codegen, not all TVM functions). However, it looks like the passes are skipping all the functions in these modules because they don't use the default compiler. Is there another way we can manually run the passes?

**Comment by zhiics** [comment]
Can you post the questions on discussion forum?

**Comment by mbaret** [comment]
Yep, I'll try and give a better explanation there in the morning.
